components/speedreader/resources/panel/components/main-panel/index.tsx

```diff
@@ -1,42 +1,127 @@
 import { FontStyleList, ContentList } from '../lists'
-import ZoomControl from '../zoom-control'
+import FontSizeControl from '../font-size-control'
 import ThemeControl from '../theme-control'
+import { panelDataHandler, SiteSettings, Theme, FontSize, ContentStyle, FontFamily } from '../../api/browser'
+import BackgroundSVG from '../../svg/background'
 
 function MainPanel () {
+  const [siteSettings, setSiteSettings] = React.useState<SiteSettings | undefined>()
+
+  React.useEffect(() => {
+    const onVisibilityChange = () => {
+      if (document.visibilityState === 'visible') {
+        panelDataHandler.getSiteSettings().then(res => setSiteSettings(res.siteSettings))
+      }
+    }
+
+    document.addEventListener('visibilitychange', onVisibilityChange)
+    panelDataHandler.getSiteSettings().then(res => setSiteSettings(res.siteSettings))
+
+    return () => {
+      document.removeEventListener('visibilitychange', onVisibilityChange)
+    }
+  }, [])
+
+  if (!siteSettings) {
+    return null
+  }
+
+  const handleThemeChange = (theme: Theme) => {
+    setSiteSettings({ ...siteSettings, theme })
+    panelDataHandler.setTheme(theme)
+  }
+
+  const handleFontSizeChange = (fontSize: FontSize) => {
+    setSiteSettings({ ...siteSettings, fontSize })
+    panelDataHandler.setFontSize(fontSize)
+  }
+
+  const handleContentStyleChange = (contentStyle: ContentStyle) => {
+    setSiteSettings({ ...siteSettings, contentStyle })
+    panelDataHandler.setContentStyle(contentStyle)
+  }
+
+  const handleFontFamilyChange = (fontFamily: FontFamily) => {
+    setSiteSettings({ ...siteSettings, fontFamily })
+    panelDataHandler.setFontFamily(fontFamily)
+  }
+
+  const handleChange = (isOn: boolean) => {
+    panelDataHandler.setEnabled(isOn)
+    panelDataHandler.getSiteSettings().then(res => setSiteSettings(res.siteSettings))
+  }
+
+  let mainContentElement = (
+    <>
+      <S.Section>
+        <div className="title">Theme</div>
+        <ThemeControl
+          activeTheme={siteSettings.theme}
+          onClick={handleThemeChange}
+        />
+      </S.Section>
+      <S.Section>
+        <div className="title">Font style</div>
+        <FontStyleList
+          activeFontFamily={siteSettings.fontFamily}
+          onClick={handleFontFamilyChange}
+        />
+        <br />
+        <FontSizeControl
+          currentSize={siteSettings.fontSize}
+          onClick={handleFontSizeChange}
+        />
+      </S.Section>
+      <S.Section>
+        <div className="title">Content</div>
+        <ContentList
+          activeContentStyle={siteSettings.contentStyle}
+          onClick={handleContentStyleChange}
+        />
+      </S.Section>
+    </>
+  )
+
+  let headerTextElement = (
+    <S.SiteTitleBox>
+      Always load speedreader mode for
+      <S.SiteName>{siteSettings.host}</S.SiteName>
+    </S.SiteTitleBox>
+  )
+
+  if (!siteSettings.isEnabled) {
+    headerTextElement = <></>
+    mainContentElement = (
```

Comment:
Seems a shame to construct all the react components on every render just to not use them and instead use a different set of components. Wouldn't it be better to always conditionally set `headerTextElement` and `mainContentElement` based on the `siteSettings.isEnabled` value, whether it's using simple conditional or ternary?

Reply:
makes sense